### PR TITLE
Add regions endpoint

### PR DESCRIPTION
**Pull request by ryanuber**
This adds an API call for fetching the list of regions known to Nomad. It also returns a list of the datacenter names nested under each. Related: #243

### CODE REVIEW

File: nomad/server.go [hunk 1]
```diff
@@ -612,6 +615,30 @@ func (s *Server) State() *state.StateStore {
...
 }
 
+// Regions returns the known regions in the cluster along with a list
+// of the datacenters within each region.
+func (s *Server) Regions() map[string][]string {
+	s.peerLock.RLock()
+	defer s.peerLock.RUnlock()
+
+	regions := make(map[string][]string, len(s.peers))
+	for region, servers := range s.peers {
```
Comment (diptanu) [inline nomad/server.go]: With this implementation, we will get only the regions where the Nomad servers are present. Users might expect the API to return the entire list of regions where the Nomad cluster spans. We might need to change the name of the API or document it to make sure people understand the purpose of the API.

File: nomad/regions_endpoint_test.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	// Query the regions list
+	testutil.WaitForResult(func() (bool, error) {
+		var arg structs.GenericRequest
+		var out []string
+		if err := msgpackrpc.CallWithCodec(codec, "Region.List", &arg, &out); err != nil {
+			t.Fatalf("err: %v", err)
+		}
+		if len(out) != 2 || out[0] != "region1" || out[1] != "region2" {
```
Comment (dadgar) [inline nomad/regions_endpoint_test.go]: Is this going to be flaky? Is there any guarantee on order?

File: nomad/server_test.go [hunk 3]
```diff
@@ -86,3 +88,34 @@ func TestServer_RPC(t *testing.T) {
...
+	s2Addr := fmt.Sprintf("127.0.0.1:%d",
+		s2.config.SerfConfig.MemberlistConfig.BindPort)
+	if n, err := s1.Join([]string{s2Addr}); err != nil || n != 1 {
+		t.Fatalf("Failed joining: %v (%d joined)", err, n)
+	}
+
+	// Try listing the regions
+	testutil.WaitForResult(func() (bool, error) {
+		out := s1.Regions()
+		if len(out) != 2 || out[0] != "region1" || out[1] != "region2" {
```
Comment (dadgar) [inline nomad/server_test.go]: Same

File: nomad/server.go [hunk 4]
```diff
@@ -612,6 +615,18 @@ func (s *Server) State() *state.StateStore {
 	return s.fsm.State()
 }
 
+// Regions returns the known regions in the cluster.
+func (s *Server) Regions() []string {
+	s.peerLock.RLock()
+	defer s.peerLock.RUnlock()
+
+	regions := make([]string, 0, len(s.peers))
```
Comment (dadgar) [inline nomad/server.go]: What if I have multiple servers per region. You probably want to make a map to keep track of unique regions

File: nomad/regions_endpoint_test.go (hunk 2 above)
Comment (ryanuber) [inline nomad/regions_endpoint_test.go]: Good call, tests didn't indicate anything but a quick `sort.Strings()` should fix this up.

File: nomad/server.go (hunk 4 above)
Comment (ryanuber) [inline nomad/server.go]: The peers map key is region ID, which would prevent having any duplicates. There might be multiple servers but they are still all nested under the same map key.

**Comment by dadgar** [comment]
LGTM

**Pull request merged**